proxy/proxy/client.go

```diff
@@ -91,6 +95,11 @@ type cacheEntry struct {
 // proxy them to the destination instance. It blocks until connSrc is closed.
 func (c *Client) Run(connSrc <-chan Conn) {
 	for conn := range connSrc {
+		if c.MaxConnections > 0 && len(c.Conns.m) >= c.MaxConnections {
```

Comment:
You need to check the number of connections in the entire map, not just the `len(c.Conns.m)` (which is the number of unique "id"s in the map); also you need to hold the lock in the ConnSet or else there's going to be a data race here. (we shouldn't read the unexported 'm' variable in this code anyway, even though it's allowed by the compiler because it happens to be in the same package.)

Even if you held the lock here or wrote a method on ConnSet to get the number of connections in the set, there's a logical race between checking the number of connections here and adding the new connection to the ConnSet in handleConn. Therefore I think this check needs to happen in handleConn.

Unfortunately I think this is too inefficient to work well anyway since you're adding an iteration through the ConnSet map on each new connection request, not to mention the inefficiency of ConnSet in the first place: I'm already hoping to remove a lot of the ConnSet logic from this Client because it is unfortunately pretty inefficient already (some `O(n)` operations which I completely forgot about and extra locking which probably reduces parallelism a little more than we'd like).

Barring the extra work to optimize the ConnSet, it's probably easiest to use a counter variable using atomic increments in this Client type to make this happen. That won't add much latency to new connections and avoids undesired coupling to the ConnSet. We can always move where that counter is in the future (it sort of belongs in the ConnSet, but would require some new methods on that type to make it work well).